### ISSUE TITLE
Wrong characters with Swedish keyboard

### ISSUE THREAD

**Issue by wilhelmloof** [pqrs-org/Karabiner-Elements#2791]
Simple modifications gives wrong characters with Swedish keyboard layout. I am trying to create a new backslash key.

*Results for after modifying a key*
To key: backslash (\) gives: '
To key: non_us_backslash gives: <

And is there a way to change so that the modified key only is evoked after pressing a meta key? I would to prefer to press Ctrl + '^' to type backslash, since Ctrl + '^' is empty and close to the location of backslash on US keyboard layout.

### System
OS: macOS Mojave ( 10.14.6), Swedish
Keyboard: Apple A1243, Swedish

**Comment by MuhammedZakir** [comment]
IIRC, Karabiner-Elements doesn't work well with other layouts. As a workaround, you can swap wrong keys with the correct one using complex modification.

> And is there a way to change so that the modified key only is evoked after pressing a meta key? [...]

It's only possible with complex modification.

https://karabiner-elements.pqrs.org/docs/json/

**Comment by wilhelmloof** [comment]
> IIRC, Karabiner-Elements doesn't work well with other layouts. As a workaround, you can swap wrong keys with the correct one using complex modification.
> 
> > And is there a way to change so that the modified key only is evoked after pressing a meta key? [...]
> 
> It's only possible with complex modification.
> 
> https://karabiner-elements.pqrs.org/docs/json/

Thanks! But that rabbit whole was deep.

I cobbled together this. But nothing happened. What is wrong?

```
{
  "title": "My Complex modifications",
  "rules": [
    {
      "description": "Backslash to Ctrl + ^ (Swedish)",
      "manipulators": [
        {
          "type": "basic",
          "from": {
            "key_code": "36",
            "modifiers": {
              "modifiers": ["left_shift", "left_option"]
            }
          },
          "to": [{
              "key_code": "48",
              "modifiers": ["left_control"]
            }]
        }
      ]
    }
  ]
}
```
Instead of 'left_control' and 'left_option' etc, is it ok to write just 'control' for both left and right?

**Comment by MuhammedZakir** [comment]
If you are saving that into `~/.config/karabiner/assets/complex_modifications/aName.json`, then that's correct. If you are directly adding that to `~/.config/karabiner/karabiner.json`, then you only need what's inside `rules`.

---

When entering raw key codes, you shouldn't quote numbers; same goes for booleans (`true`/`false`).

E.g. `"key_code": 22`

Raw key number will work, but I'm not whether it accepts hexadecimal values though. You might have to convert it to decimal.

**Comment by wilhelmloof** [comment]
Great

Both these worked `"key_code": "7"` or `"key_code": 36`.

But it seems like Karabiner doesn't work on an enough low level. My basic problem is that Google Sheets has short keys that you can't change, and they interfere with normal usage on a non US keyboard. Like I wan't to make a backslash character, that is Alt+Shift+7 on my keyboard, but Sheets interpets that as short key for making a black border around the cell. The problem remains after making the complex key modification, Karabiner still "presses the underlying keys" when the user presses the new key combo. That's the whole thing with virtual keyboards. I need some lower level of key remapping. What software can do this on macOS? [Ukelele](https://software.sil.org/ukelele/)?

**Comment by MuhammedZakir** [comment]
Karabiner-Elements works at the lowest level. You can't go any lower than that (at least there is no "official" way).

---

I understand what the problem is. I can't give you an answer before testing (if you can't wait, see the next section). Also, I think you are misunderstanding something here. Karabiner-Elements map a key or a group of keys to either some other key(s), or a shell command, or a NOP key.

In the above config, you are remapping `Control+]` to `Alt+Shift+7`, and not `backslash` per se. It will be as if you have pressed `Alt+Shift+7` instead of `Control+]`. How an app, or the system interprets it is beyond Karabiner.

---

If you want a hacky workaround, you can just literally insert a `backslash` using `to.shell_command`. Either use AppleScript,
```
tell application "System Events" to keystroke "\\\\"
```
or shell command (copy backslash and paste it):
```
printf "\\\\" | pbcopy; pbpaste
```

---

About Ukelele: I don't know as I haven't used it. You may give it a try.

**Comment by wilhelmloof** [comment]
Ok, that is promising.

But how do can I print a list of all key_code:s and what the represent in ANSI with a ISO keyboard? This [list](https://www.fileformat.info/info/unicode/char/005c/index.htm) states that 92 i backslash in decimal. But that Karabiner gives '4'

**Comment by MuhammedZakir** [comment]
> But how can I print a list of all key_code:s and what the represent in ANSI with a ISO keyboard? This [list](https://www.fileformat.info/info/unicode/char/005c/index.htm) states that 92 i backslash in decimal. But 92 in Karabiner gives '4'.

The link you mentioned is for Unicode values, and not keyboard's key codes.

92 is the key code for button keypad4. You can find full list here [1][2].

[1] https://opensource.apple.com/source/IOHIDFamily/IOHIDFamily-315.7.16/IOHIDFamily/IOHIDUsageTables.h
    - See `KeyboardOrKeypad Page (0x07)` enum -- values are in hex. 
[2] https://opensource.apple.com/source/IOHIDFamily/IOHIDFamily-700/IOHIDFamily/AppleHIDUsageTables.h.auto.html
    - Apple-specific keycodes such as `Launchpad`, `Expose`, `Spotlight`, `Fn`, etc.

---

> And what is Apple's keyboard country code for Sweden? I can't find any info at Apple Dev.

I don't know. Is country code useful?

> ```
> $ defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"
> {
>     "10203-5824-0" = 43;
> }
> ```
> 
> What is 43 in that output?

I don't know. Probably type of the keyboard - ISO, ANSI or JIS.

**Comment by wilhelmloof** [comment]
Thanks for hanging in and helping me @MuhammedZakir 

I had to set a number higher than zero in Virtual Keyboard in order to get som keys to work on ISO layout.

The problem is that I can't find a way to produce backslash without pressing '7' with non of the key_codes I've tested.
There is a missmatch between input and output with my non US keyboard and Karabiner. Example 'backslash' with key code 49 produces '’' and 'non_us_backslash' key code 100 produces '<'. Is there some way that I can write the hex value directly as a key_code?

I also tried [Key Codes app](https://manytricks.com/keycodes/) and it reported backslash as key code 26, but that gives 'w' with Karabiner. 

This was the output from Key Codes app for backslash
Key Code: 26 / 0x1a
Modifiers: 655650 / 0xa0122

I also tested adding that modifier in the Karabiner json:
```
          "to": [{
              "key_code": 26,
              "modifiers": [655650]
            }]
```
I got nothing.

**Comment by MuhammedZakir** [comment]
Before going further, I have a doubt/question. Why do you need to press `Alt+Shift+7` to get `backslash`? Does you keyboard not have a `backslash` key? Either by pressing directly or while holding `Shift`? If you have one, I recommend you to insert a `backslash` and use whatever `key_code` or name Karabiner reports. Unless you're using an ANSI keyboard (which you're not), or you're familiar with ISO keyboard's key codes, it's better to not to go too deep into "key codes" stuff.

> [...] Is there some way that I can write the hex value directly as a key_code?

Have you tried using hex values in `key_code`? I haven't, but it might work.

E.g. `"key_code": 0x1A` (for `w`).

---

> There is a missmatch between input and output with my non US keyboard and Karabiner. 

This is not Karabiner's fault AFAIK. All key codes used/reported by Karabiner (or any other app which uses similar method) uses ANSI layout even if the keyboard uses a different layout. From the link I posted earlier:

> Note: A general note on Usages and languages: Due to the variation of keyboards from language to language, it is not feasible to specify exact key mappings for every language. Where this list is not specific for a key function in a language, the closest equivalent key position should be used, so that a keyboard may be modified for a different language by simply printing different keycaps. One example is the Y key on a North American keyboard. In Germany this is typically Z. Rather than changing the keyboard firmware to put the Z Usage into that place in the descriptor list, the vendor should use the Y Usage on both the North American and German keyboards. This continues to be the existing practice in the industry, in order to minimize the number of changes to the electronics to accommodate otherlanguages.

---

The key codes reported by KeyCodes app are different from the ones used by Karabiner. Both of them use different methods for capturing keys - Karabiner uses `IOKit` [1], whereas, IIAR, KeyCodes uses `CGEvent`. You can use the key codes reported by latter if you use `CGEvent`, or for AppleScript (I think). That's why, you got `w` for key code `26`(`0x1A`) (check the link I posted earlier).

[1] https://github.com/pqrs-org/Karabiner-Elements/blob/main/docs/DEVELOPMENT.md#the-difference-of-event-grabbing-methods

**Comment by wilhelmloof** [comment]
No, Nordic keyboard layout does not have a backslash key, that's the main problem.

Thanks for clarifying the different key codes from IOKit and CGEvent.

I can't find any documentation about writing hex codes in Karabiner JSON.

**Comment by MuhammedZakir** [comment]
> No, Nordic keyboard layout does not have a backslash key, that's the main problem.

I see. One more question: when you mention Nordic keyboard layout, are you referring to the keyboard itself, or the input layout? What is the output of EventViewer when you insert a `backslash`?
 
> Thanks for clarifying the different key codes from IOKit and CGEvent.

You're welcome! :-)

> I can't find any documentation about writing hex codes in Karabiner JSON.

Try inserting like you would for a decimal number: `"key_code": 0x4`.

**Comment by wilhelmloof** [comment]
Karabiner didn't accept this `"shell_command": "printf "\\\\" | pbcopy; pbpaste"`, this worked in terminal but not in Karabiner               `"shell_command": "printf '\\' | pbcopy; pbpaste"`.

**Comment by MuhammedZakir** [comment]
1. You didn't escape `"` inside command
```
"shell_command": "printf \"\\\\\" | pbcopy; pbpaste"
```

2. For the second and third command, you didn't escape backslashes. Your executed command will then become `printf '\' | pbcopy; pbpaste`, which as you can see is wrong. You need to escape `\\` two times here - one for JSON, and one for shell.
```
"shell_command": "printf '\\\\' | pbcopy; pbpaste"
```

---

Another solution is to create a custom layout (Ukelele?), in which, `backslash` can be accessed directly, or while holding shift.

**Comment by MuhammedZakir** [comment]
Another (slightly convoluted) way is, when pressing `Alt+Shift+7`, switch input source [1] to US layout, enter `backslash` and switch back to Swedish layout.

[1] https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to/select-input-source/

**Comment by wilhelmloof** [comment]
Escape twice! One for JSON and one for the shell haha, of course, this is getting even hackier 😄 
Karabiner accepted both `"shell_command": "printf \"\\\\\" | pbcopy; pbpaste"` and `"shell_command": "printf '\\\\' | pbcopy; pbpaste"`, but non of them produced any output.

Switching input language, maybe, but unfortunately Swedish keyboard layout doesn't have a key in the position os US layout. That position is occupied by a larger enter key.

**Comment by MuhammedZakir** [comment]
My fault! I forgot about one thing. `pbpaste` prints the clipboard conten, and doesn't paste clipboard content at the current cursor location. For that we will have to copy `backslash` and then invoke `command+v`:
```json
{
  "description": "Alt+Shift+7 => backslash",
  "manipulators": [{
    "type": "basic",
    "from": {
      "key_code": "7",
      "modifiers": {
        "mandatory": ["shift", "option"],
        "optional": ["caps_lock"]
      }
    },
    "to": [{
      "shell_command": "printf '\\\\' | pbcopy"
    }],
    "to_delayed_action": {
      "to_if_invoked": [{
        "key_code": "v",
        "modifiers": "command"
      }]
    },
    "parameters": {
      "basic.to_delayed_action_delay_milliseconds": 50
    }
  }]
}
```

Note: You can increase delay if 50ms is too fast. You may also remove `modifiers.optional` if you don't want this to work when caps lock is enabled. This is helpful if you want to use `Alt+Shift+7`, so you can use `caps_lock` as a toggler -- inserting backslash vs pressing `Alt+Shift+7`.

---

> Switching input language, maybe, but unfortunately Swedish keyboard layout doesn't have a key in the backslash position of US layout. That position is occupied by a larger enter key.

Switch to US input, and then find which key corresponds to `backslash`. Then, you can use that key for `backslash`. You can use on-screen keyboard for testing. I have tested inserting another character by switching input layout. It works nicely. You can use the above method in case you wanna try.

**Comment by wilhelmloof** [comment]
Thanks! No `pbpaste` seems logical. But strangely Karabiner did not produce any output with the shell hack. I used this code:

```
{
    "description": "Alt+Shift+7 => backslash",
    "manipulators": [
        {
            "type": "basic",
            "from": {
                "key_code": "close_bracket",
                "modifiers": {
                    "mandatory": [
                        "control"
                    ]
                }
            },
            "to": [
                {
                    "shell_command": "printf '\\\\' | pbcopy"
                }
            ],
            "to_delayed_action": {
                "to_if_invoked": [
                    {
                        "key_code": "v",
                        "modifiers": "command"
                    }
                ]
            },
            "parameters": {
                "basic.to_delayed_action_delay_milliseconds": 50
            }
        }
    ]
}
```

It might be better after all to switch to ANSI or US layout. Backslash lands on `"key_code": "non_us_pound"` without modifiers when setting input to ANSI with my ISO keyboard.

How to write switching between layouts? Is setting ANSI enough? And does the JSON data also need to have another statement that switches back to ISO?

**Comment by wilhelmloof** [comment]
This attempt with switching to ANSI didn't work:

```
{
    "description": "Ctrl caret >> backslash",
    "manipulators": [
        {
            "type": "basic",
            "from": {
                "input_source_id": "^com\\.apple\\.keylayout\\.Swedish-Pro$",
                "key_code": "close_bracket",
                "modifiers": {
                    "mandatory": [
                        "control"
                    ]
                }
            },
            "to": [
                {
                    "input_source_id": "^com\\.apple\\.keylayout\\.US$",
                    "key_code": "non_us_pound"
                }
            ]
        }
    ]
}
```

**Comment by MuhammedZakir** [comment]
> Thanks! Not using `pbpaste` seems to be correct. But strangely Karabiner didn't produce any output with the shell hack. I used this code:
> 
> ```
> --snip--
> ```

The config I posted worked for me. However, the one I posted was "handwritten" and there might be some errors.

1. Check if there is any error in the config - `Karabiner-Elements Preferences > Log`.
2. If there is no error, then the reason is probably the delay is not enough. You can test this by verifying that the backslash is indeed copied to the clipboard after pressing `control+caret` by invoking paste manually (`command+v`, or right-click context menu, or `pbcopy`) .
    - If you get backslash when you paste, then the problem is with delay.
    - if you don't get backslash, then the problem is with the shell command. But, `printf '\\\\' | pbcopy` does work for me.

---

> It might be better after all to switch to ANSI or US layout. Backslash lands on `"key_code": "non_us_pound"` without modifiers when setting input to ANSI with my ISO keyboard.
> 
> How to write switching between layouts? Is setting ANSI enough? And does the JSON data also need to have another statement that switches back to ISO?

> This attempt with switching to ANSI didn't work:
> 
> ```
> --snip--
> ```

I have tested the config, and it's working for me. But, this is also "handwritten", so check if there is any errors. And, increase `hold_down_milliseconds` if it's not enough.

```json
{
  "description": "Control+Caret => backslash",
  "manipulators": [{
    "type": "basic",
    "from": {
      "key_code": "close_bracket",
      "modifiers": {
        "mandatory": [ "control" ]
      }
    },
    "to": [
      {
        "select_input_source": {
          "language": "^en$"
        }
      },
      {
        "key_code": "vk_none",
        "hold_down_milliseconds": 80
      },
      {
        "key_code": "non_us_pound"
      }
    ],
    "to_after_key_up": [{
      "select_input_source": {
        "input_source_id": "^com.apple.keylayout.Swedish-Pro$"
      }
    }]
  }]
}
```

**Comment by MuhammedZakir** [comment]
P.S. You can restrict that modfication to Swedish keylayout using `input_source_if` [1].

[1] https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/conditions/input-source/

**Comment by wilhelmloof** [comment]
Thank you!! This seems to be correct. Now it works as desired. And now I can write backslashes in Google Sheets 😄 

The reason that some of the code worked for you and not for me was that the wrapper was missing. This part:
```
{
    "title": "Complex modifications",
    "rules": [
           … modifications …
    ]
}
```

I was a bit confused that in the` "select_input_source"` part only language is specified, but in the `"to_after_key_up"` `"input_source_id"` is set. Setting language in `"to_after_key_up"` might be more logical, but some languages has several keyboard layouts. Then it could be safer to set layout. I am not criticizing your solution, I just want to understand how Karabiner works.

What is `"vk_none"`? 
Why is `"hold_down_milliseconds"` needed for this? In the variant with `"shell_command"` it seemed fitting, but when the shell is not used?

(Note: first I wrote the JSON in TextMate and Smultron, but then VS Codes. VSC has auto validation of JSON, directly finds matching brackets etc.)

**Comment by MuhammedZakir** [comment]
> Thank you!! This seems to be correct. Now it works as desired. And now I can write backslashes in Google Sheets 

Glad to hear! :-D

> The reason that some of the code worked for you and not for me was that the wrapper was missing. This part:
> 
> ```
> {
>     "title": "Complex modifications",
>     "rules": [
>            … modifications …
>     ]
> }
> ```

Ah! In case, you are still confused, see https://karabiner-elements.pqrs.org/docs/json/root-data-structure/.

> I was a bit confused that in the `"select_input_source"` part only language is specified, but in the `"to_after_key_up"` `"input_source_id"` was set. Setting language in `"to_after_key_up"` might look more logical, but some languages has several keyboard layouts. Then it could be safer to set layout. I am not criticizing your solution, I just want to understand how Karabiner works.

The reason is exaclty what you said. We just want to insert backslash, we don't care which English layout is used. That's why, I used `language` in `select_input_source`. Whereas, in `to_after_key_up`, I used `input_source_id` because you are using Swedish-Pro layout. If we only use `language`, then the layout selected is (IIRC) Swedish layout. If you don't care about it, you can use `language`.

tl;dr - use the one which suites your need.

> https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/conditions/keyboard-type/

?

> What is `"vk_none"`?

This key code is used to "disable" keys and can only be used in `to` event. To main uses of `vk_none` are for ignoring stuck keys in your keyboard, and also for ignore certain keypresses in certain layouts or emulation modes.

> Why is `"hold_down_milliseconds"` needed for this? In the variant with `"shell_command"` it seemed fitting, but when the shell is not used?

Actually, `hold_down_milliseconds` were used for modifier keys such as `caps_lock` and `shift`. With it, you can create a temporary sticky key which turns off after a specified time. In this case, we are using it to create a short delay between switching input and emitting `non_us_pound`. Without it, `non_us_pound` will be emitted before the layout get switched.

**Comment by wilhelmloof** [comment]
Thanks for your explanations.

> ?

I was linking to the documentation about ANSI and ISO. But it's probably safer to set Swedish-Pro layout instead of ANSI or `sv`. My idea about just switching between ANSI and ISO, was that the modification would not be layout or language specific and thus easier to use on other machines without being edited.

> Actually, `hold_down_milliseconds` are used for modifier keys such as `caps_lock` and `shift`. With it, you can create a temporary sticky key which turns off after a specified time. In this case, we are using it to create a short delay between switching input and emitting `non_us_pound`. Without it, `non_us_pound` will be emitted before the layout get switched. You can use any modifier key instead of `vk_none`. I used `vk_none` here just to be safe, and FWIW, I used `shift` when I first tested this,

Ok. So switching layout is not fast enough (or has to be triggered) so you are using a blank key. Smart

Karabiner is powerful but the syntax is tricky and a bit annoying that non US layouts has to declare 'wrong' characters in order to get the desired characters.

**Comment by MuhammedZakir** [comment]
> Thanks for your explanations.

You're welcome! :-)

> > ?
> 
> I was linking to the documentation about ANSI and ISO. But it's probably safer to set Swedish-Pro layout instead of ANSI or `sv`. My idea about just switching between ANSI and ISO, was that the modification would not be layout or language specific and thus easier to use on other machines without being edited.

That could work. But in your case, there is no backslash which is why you had to do this, whereas that may not be the case with other ISO keyboards. This modification may not be needed in such keyboards.
 
> Ok. So switching layout is not fast enough (or has to be triggered) so you are using a blank key. Smart

Yes.
 
> Karabiner is powerful, but the syntax is tricky

There is a proposal to use TOML, or other simpler syntax. JSON is not suited for this, especially for big config.

> and a bit annoying that non US layouts has to declare 'wrong' characters in order to get the desired characters. Other than that, Karabiner does its job well.

This is what Apple says:
```
Note: A general note on Usages and languages: Due to the variation of keyboards from language to language, it is not feasible to specify exact key mappings for every language. Where this list is not specific for a key function in a language, the closest equivalent key position should be used, so that a keyboard may be modified for a different language by simply printing different keycaps. One example is the Y key on a North American keyboard. In Germany this is typically Z. Rather than changing the keyboard firmware to put the Z Usage into that place in the descriptor list, the vendor should use the Y Usage on both the North American and German keyboards. This continues to be the existing practice in the industry, in order to minimize the number of changes to the electronics to accommodate otherlanguages.
```

Karabiner could show key codes corresponding to the ISO keybaord to users, and iternally map them to "correct" key codes. However, with different variations and layouts out there, I am not sure how feasible that would be, and tekezo alone has to maintain all these which would be too much work and not worth it IMHO.

---

If this problem is solved, please close this. If you have other questions/problems, feel free to open a new issue.
